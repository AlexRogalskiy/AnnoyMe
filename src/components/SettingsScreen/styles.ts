import { StyleSheet } from "react-native";
import * as colors from "../../colors";

export default StyleSheet.create({
  container: {
    flex: 1,
    padding: 20,
    justifyContent: "flex-start",
  },
  inputContainer: {
    paddingBottom: 10,
  },
  inputTitle: {
    fontWeight: "200",
  },
  input: {
    borderBottomWidth: 1,
  },
  inputValid: {
    borderBottomColor: colors.green200,
  },
  inputInvalid: {
    borderBottomColor: colors.red500,
  },
});
