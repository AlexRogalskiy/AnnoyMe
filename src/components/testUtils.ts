import { act } from "react-test-renderer";
import Settings from "../dto/Settings";

/** Waits for useEffect effects. */
export const waitForEffects = () =>
  new Promise(resolve => setTimeout(resolve, 0));

/** Creates mock of settings repository. `resolve` should be called if settings value will be used. */
export const makeSettingsRepository = (settings?: Settings): any => {
  let resolve: (value: Settings) => void | undefined;
  const settingsPromise = new Promise(_resolve => (resolve = _resolve));

  return {
    read: jest.fn(() => settingsPromise),
    save: jest.fn((_: Settings) => undefined),

    resolve: async () => {
      act(() => resolve!(settings!));

      await waitForEffects();
    },
  };
};
