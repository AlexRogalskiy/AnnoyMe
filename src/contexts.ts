import React from "react";
import ITasksRepository from "./services/tasksRepository/ITasksReposiotry";
import ITaskValidator from "./services/taskValidator/ITaskValidator";
import ISettingsRepository from "./services/settingsRepository/ISettingsRepository";
import ISettingsValidator from "./services/settingsValidator/ISettingsValidator";

interface ServicesContextProps {
  tasksRepository: ITasksRepository;
  taskValidator: ITaskValidator;
  settingsRepository: ISettingsRepository;
  settingsValidator: ISettingsValidator;
}

export const ServicesContext = React.createContext<ServicesContextProps>(
  {} as ServicesContextProps,
);

interface ConfigurationContextProps {
  refreshInterval: number;
}

export const ConfigurationContext = React.createContext<
  ConfigurationContextProps
>({} as ConfigurationContextProps);
