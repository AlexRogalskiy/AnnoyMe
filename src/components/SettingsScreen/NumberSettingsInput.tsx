import React from "react";
import { Text, TextInput, View } from "react-native";
import styles from "./styles";

interface Props {
  label: string;
  value: number;

  validate: (value: number) => boolean;
  onChange: (value: number) => void;

  testID?: string;
}

/** Input for numbers with validation. */
export default ({ label, value, validate, onChange }: Props) => (
  <View style={styles.inputContainer}>
    <Text style={styles.inputTitle}>{label}</Text>
    <TextInput
      style={[
        styles.input,
        validate(Number(value)) ? styles.inputValid : styles.inputInvalid,
      ]}
      defaultValue={(value || "").toString()}
      onChangeText={newValue => onChange(Number(newValue))}
      keyboardType="number-pad"
      testID="number-settings-input-text-input"
    />
  </View>
);
