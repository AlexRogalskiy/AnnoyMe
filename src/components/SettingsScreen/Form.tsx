import React from "react";
import { View } from "react-native";
import * as constants from "../../constants";
import NumberSettingsInput from "./NumberSettingsInput";
import styles from "./styles";

const isValidHour = (hour: number) => hour >= 0 && hour <= 24;

const isValidFrequency = (minutes: number) =>
  minutes >= constants.MIN_FREQUENCY && minutes <= 120;

interface Props {
  startHour: number;
  endHour: number;
  frequency: number;

  onNumberSettingsChange: (key: string, value: number) => void;
}

export default ({
  startHour,
  endHour,
  frequency,
  onNumberSettingsChange,
}: Props) => (
  <View style={styles.container}>
    <NumberSettingsInput
      label="Annoy from hour"
      value={startHour}
      validate={isValidHour}
      onValidChange={val => onNumberSettingsChange("startHour", val)}
    />
    <NumberSettingsInput
      label="Annoy till hour"
      value={endHour}
      validate={isValidHour}
      onValidChange={val => onNumberSettingsChange("endHour", val)}
    />
    <NumberSettingsInput
      label="Annoys frequency"
      value={frequency}
      validate={isValidFrequency}
      onValidChange={val => onNumberSettingsChange("frequency", val)}
    />
  </View>
);
