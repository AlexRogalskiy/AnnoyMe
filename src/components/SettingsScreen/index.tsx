import React, { Suspense } from "react";
import { editableSettings } from "../../hooks/settings";
import Loading from "../Loading";
import Form from "./Form";

const SettingsScreen = () => {
  const { settings, changeSetting } = editableSettings();

  return (
    <Suspense fallback={<Loading />}>
      {settings && <Form onSettingChange={changeSetting} {...settings} />}
    </Suspense>
  );
};

SettingsScreen.navigationOptions = {
  title: "Settings",
};

export default SettingsScreen;
