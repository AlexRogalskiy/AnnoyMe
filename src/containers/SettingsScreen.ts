import { bindActionCreators, Dispatch, AnyAction } from "redux";
import { connect } from "react-redux";
import {
  setStartHour,
  setEndHour,
  setFrequency,
} from "../store/settings/actions";
import { State } from "../store";
import SettingsScreen from "../components/SettingsScreen";

const mapStateToProps = (state: State) => ({
  startHour: state.settings.startHour,
  endHour: state.settings.endHour,
  frequency: state.settings.frequency,
});

const mapDispatchToProps = (dispatch: Dispatch<AnyAction>) =>
  bindActionCreators(
    {
      setStartHour,
      setEndHour,
      setFrequency,
    },
    dispatch,
  );

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(SettingsScreen);
