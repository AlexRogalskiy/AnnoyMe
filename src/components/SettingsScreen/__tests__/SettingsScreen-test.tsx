import { last } from "lodash";
import React from "react";
import renderer, { act } from "react-test-renderer";
import { ServicesContext } from "../../../contexts";
import { makeSettingsRepository, waitForEffects } from "../../testUtils";
import SettingsScreen from "../SettingsScreen";

const settings = {
  startHour: 10,
  endHour: 22,
  frequency: 15,
};

it("SettingsScreen renders loading screen when settings not ready", () => {
  const settingsRepository = makeSettingsRepository();

  const tree = renderer
    .create(
      <ServicesContext.Provider value={{ settingsRepository } as any}>
        <SettingsScreen />
      </ServicesContext.Provider>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot("SettingsScreen-loading");
});

it("SettingsScreen renders settings form", async () => {
  const settingsRepository = makeSettingsRepository(settings);

  const screenRenderer = renderer.create(
    <ServicesContext.Provider value={{ settingsRepository } as any}>
      <SettingsScreen />
    </ServicesContext.Provider>,
  );

  await settingsRepository.resolve();

  const tree = screenRenderer.toJSON();
  expect(tree).toMatchSnapshot("SettingsScreen-form");
});

it("SettingsScreen saves valid changes", async () => {
  const settingsRepository = makeSettingsRepository(settings);

  const screenRenderer = renderer.create(
    <ServicesContext.Provider value={{ settingsRepository } as any}>
      <SettingsScreen />
    </ServicesContext.Provider>,
  );

  await settingsRepository.resolve();

  const newSettings = {
    startHour: 12,
    endHour: 23,
    frequency: 120,
  };

  const form = screenRenderer.root.findByProps({ testID: "settings-form" });
  act(() => form.props.onChange(newSettings));

  await waitForEffects();

  expect(last(settingsRepository.save.mock.calls)).toEqual([newSettings]);
});

it("SettingsScreen ignores invalid changes", async () => {
  const settingsRepository = makeSettingsRepository(settings);

  const screenRenderer = renderer.create(
    <ServicesContext.Provider value={{ settingsRepository } as any}>
      <SettingsScreen />
    </ServicesContext.Provider>,
  );

  await settingsRepository.resolve();

  const newSettings = {
    startHour: -12,
    endHour: -23,
    frequency: -120,
  };

  const form = screenRenderer.root.findByProps({ testID: "settings-form" });
  act(() => form.props.onChange(newSettings));

  await waitForEffects();

  expect(last(settingsRepository.save.mock.calls)).not.toEqual([newSettings]);
});
