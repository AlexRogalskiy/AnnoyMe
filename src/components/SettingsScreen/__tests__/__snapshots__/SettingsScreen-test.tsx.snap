// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsScreen renders loading screen when settings not ready: SettingsScreen-loading 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 14,
        "padding": 20,
      }
    }
  >
    Loading...
  </Text>
</View>
`;

exports[`SettingsScreen renders settings form: SettingsScreen-form 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "flex-start",
      "padding": 20,
    }
  }
>
  <View
    style={
      Object {
        "paddingBottom": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "fontWeight": "200",
        }
      }
    >
      Annoy from hour
    </Text>
    <TextInput
      allowFontScaling={true}
      defaultValue="10"
      keyboardType="number-pad"
      onChangeText={[Function]}
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "borderBottomWidth": 1,
          },
          Object {
            "borderBottomColor": "#c8e6c9",
          },
        ]
      }
      testID="number-settings-input-text-input"
      underlineColorAndroid="transparent"
    />
  </View>
  <View
    style={
      Object {
        "paddingBottom": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "fontWeight": "200",
        }
      }
    >
      Annoy till hour
    </Text>
    <TextInput
      allowFontScaling={true}
      defaultValue="22"
      keyboardType="number-pad"
      onChangeText={[Function]}
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "borderBottomWidth": 1,
          },
          Object {
            "borderBottomColor": "#c8e6c9",
          },
        ]
      }
      testID="number-settings-input-text-input"
      underlineColorAndroid="transparent"
    />
  </View>
  <View
    style={
      Object {
        "paddingBottom": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "fontWeight": "200",
        }
      }
    >
      Annoys frequency
    </Text>
    <TextInput
      allowFontScaling={true}
      defaultValue="15"
      keyboardType="number-pad"
      onChangeText={[Function]}
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "borderBottomWidth": 1,
          },
          Object {
            "borderBottomColor": "#c8e6c9",
          },
        ]
      }
      testID="number-settings-input-text-input"
      underlineColorAndroid="transparent"
    />
  </View>
</View>
`;
